src/keyDerivation_test.c

```diff
@@ -0,0 +1,141 @@
+#include "keyDerivation.h"
+#include "test_utils.h"
+#include "hex_utils.h"
+#include "stream.h"
+#include <string.h>
+#include "errors.h"
+
+#define ARRAYLEN(array) (sizeof(array) / sizeof(array[0]))
+
+void testPrivateKeyDerivation()
+{
+#define TESTCASE(path_, pathLength_, expected_) \
+	{ \
+		uint8_t chainCode[32]; \
+		cx_ecfp_256_extended_private_key_t privateKey; \
+		derivePrivateKey(path_, pathLength_, chainCode, &privateKey); \
+		stream_t s; \
+		stream_init(&s); \
+		stream_appendFromHexString(&s, expected_); \
+		const uint8_t* expectedBuffer = stream_head(&s); \
+		EXPECT_EQ_BYTES(expectedBuffer, privateKey.d, 64); \
+	}
+
+	{
+		uint32_t path1[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 1815, HARDENED_BIP32 + 1};
+		char* expected1 = "5878bdf06af259f1419324680c4d4ce05b4343895b697144d874749925e69d41"
+		                  "648e7e50061809207f81dac36b73cd8fab2325d49d26bc65ebbb696ea59c45b8";
+
+		TESTCASE(path1, ARRAYLEN(path1), expected1);
+
+		uint32_t path2[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 1815, HARDENED_BIP32 + 1, HARDENED_BIP32 + 55 };
+		char* expected2 = "38996766f3a49d987f37b694d170eef5866397f71fe0c50108f4b3db27e69d41"
+		                  "54a57702a5bf1fde836fa243f7009e665e4e902a70699ee5e82187afc425cc6a";
+
+		TESTCASE(path2, ARRAYLEN(path2), expected2);
+
+		uint32_t path3[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 1815, HARDENED_BIP32 + 1, 31 };
+		char* expected3 = "0058ea2c7e34f781d12b0cee8d8a5aff4a28436b26137e7f8bcc1fc42ae69d41"
+		                  "0aaccec99effc497b55fbd3c4ce53a3484a34e6e4b744975f9a415c1c82ecdc3";
+
+		TESTCASE(path3, ARRAYLEN(path3), expected3);
+
+		uint32_t path4[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 1815, HARDENED_BIP32 + 1, HARDENED_BIP32 + 31, 25, HARDENED_BIP32 + 44 };
+		char* expected4 = "8038f4cd293028cca6eadf4d282bedbb6ed0e7d2e4c10128b5cb621f30e69d41"
+		                  "ff4a4b14953d542c2350e0d580a224ca628577b619351df3508ed42326343a4a";
+
+		TESTCASE(path4, ARRAYLEN(path4), expected4);
+	} {
+		uint8_t chainCode[32];
+		cx_ecfp_256_extended_private_key_t privateKey;
+
+		uint32_t tooShortPath[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 1815 };
+		EXPECT_THROWS(derivePrivateKey(tooShortPath, ARRAYLEN(tooShortPath), chainCode, &privateKey), ERR_INVALID_REQUEST_PARAMETERS);
+
+		uint32_t invalidPath1[] = { HARDENED_BIP32 + 44, 1815, HARDENED_BIP32 + 1 };
+		EXPECT_THROWS(derivePrivateKey(invalidPath1, ARRAYLEN(invalidPath1), chainCode, &privateKey), ERR_INVALID_REQUEST_PARAMETERS);
+
+		uint32_t invalidPath2[] = { HARDENED_BIP32 + 44, HARDENED_BIP32 + 33, HARDENED_BIP32 + 1 };
+		EXPECT_THROWS(derivePrivateKey(invalidPath2, ARRAYLEN(invalidPath2), chainCode, &privateKey), ERR_INVALID_REQUEST_PARAMETERS);
+	}
+#undef TESTCASE
+}
+
+void testPublicKeyDerivation()
```

Comment:
I wonder whether we should move this to a single test if we anyway test same bip paths and won't test deriving public key directly from private